File: apps/uploader/src/uploader/domain/models/wsi.py
from __future__ import annotations

from dataclasses import dataclass

from uploader.domain.models.sequencing import SamplePreparation


@dataclass(frozen=True)
class WholeSlideImaging:
    wsi_identifier: str | None = None
    belongs_to_imaging_study: str | None = None
    dicom_images_count: int | None = None
    series_start_date: str | None = None
    body_region: str | None = None
    imaging_device: str | None = None
    manufacturer_of_imaging_device: str | None = None
    software_version: str | None = None
    z_stacking: str | None = None
    objective_lens_magnification: int | None = None
    illumination_method: str | None = None
    illumination_wavelength: int | None = None
    scanning_operation_mode: str | None = None
    tissue_scan_area: int | None = None
    number_of_focal_planes: int | None = None
    distance_between_focal_planes: int | None = None
    pyramid_levels: int | None = None
    colour_icc_profile: str | None = None
    preview_available: bool | None = None
    label_available: bool | None = None
    source_assay: str | None = None
    file_format: str | None = None
    file_size: int | None = None
    image_width: int | None = None
    image_height: int | None = None
    image_depth: int | None = None
    number_of_channels: int | None = None
    channel_resolution: int | None = None
    compression_method: str | None = None
    compression_quality_label: str | None = None
    annotations_available: bool | None = None


@dataclass(frozen=True)
class SlidePreparationAssay:
    assay_identifier: str | None = None
    has_input_slide_container: str | None = None
    staining_method: str | None = None
    assay_type: str | None = None
    whole_slide_imaging: WholeSlideImaging | None = None


@dataclass(frozen=True)
class SlideContainer:
    slide_container_identifier: str | None = None
    source_fixed_block: str | None = None
    container_type: str | None = None
    section_thickness: int | None = None
    cell_type: list[str] | None = None
    tissue_type: list[str] | None = None
    slide_preparation_assay: SlidePreparationAssay | None = None


@dataclass(frozen=True)
class FixedBlock:
    block_identifier: str | None = None
    source_material: str | None = None
    name_of_fixative: str | None = None
    embedding_medium: str | None = None
    sample_preparation: SamplePreparation | None = None
    slide_container: SlideContainer | None = None


@dataclass(frozen=True)
class WsiData:
    bioptic_number: str
    source_id: str
    fixed_block: FixedBlock | None = None

File: src/Uploader/Uploader.Domain/Wsi.cs
using Uploader.Domain.Common;

namespace Uploader.Domain;

/// <summary>FAIR Genomes <c>WholeSlideImaging</c> value object.</summary>
public sealed record WholeSlideImaging : ValueObject
{
    public string? WsiIdentifier { get; init; }
    public string? BelongsToImagingStudy { get; init; }
    public int? DicomImagesCount { get; init; }
    public string? SeriesStartDate { get; init; }
    public string? BodyRegion { get; init; }
    public string? ImagingDevice { get; init; }
    public string? ManufacturerOfImagingDevice { get; init; }
    public string? SoftwareVersion { get; init; }
    public string? ZStacking { get; init; }
    public int? ObjectiveLensMagnification { get; init; }
    public string? IlluminationMethod { get; init; }
    public int? IlluminationWavelength { get; init; }
    public string? ScanningOperationMode { get; init; }
    public int? TissueScanArea { get; init; }
    public int? NumberOfFocalPlanes { get; init; }
    public int? DistanceBetweenFocalPlanes { get; init; }
    public int? PyramidLevels { get; init; }
    public string? ColourIccProfile { get; init; }
    public bool? PreviewAvailable { get; init; }
    public bool? LabelAvailable { get; init; }
    public string? SourceAssay { get; init; }
    public string? FileFormat { get; init; }
    public int? FileSize { get; init; }
    public int? ImageWidth { get; init; }
    public int? ImageHeight { get; init; }
    public int? ImageDepth { get; init; }
    public int? NumberOfChannels { get; init; }
    public int? ChannelResolution { get; init; }
    public string? CompressionMethod { get; init; }
    public string? CompressionQualityLabel { get; init; }
    public bool? AnnotationsAvailable { get; init; }
}

/// <summary>FAIR Genomes <c>SlidePreparationAssay</c> value object.</summary>
public sealed record SlidePreparationAssay : ValueObject
{
    public string? AssayIdentifier { get; init; }
    public string? HasInputSlideContainer { get; init; }
    public string? StainingMethod { get; init; }
    public string? AssayType { get; init; }
    public WholeSlideImaging? WholeSlideImaging { get; init; }
}

/// <summary>FAIR Genomes <c>SlideContainer</c> value object.</summary>
public sealed record SlideContainer : ValueObject
{
    public string? SlideContainerIdentifier { get; init; }
    public string? SourceFixedBlock { get; init; }
    public string? ContainerType { get; init; }
    public int? SectionThickness { get; init; }
    public IReadOnlyList<string>? CellType { get; init; }
    public IReadOnlyList<string>? TissueType { get; init; }
    public SlidePreparationAssay? SlidePreparationAssay { get; init; }
}

/// <summary>FAIR Genomes <c>FixedBlock</c> value object.</summary>
public sealed record FixedBlock : ValueObject
{
    public string? BlockIdentifier { get; init; }
    public string? SourceMaterial { get; init; }
    public string? NameOfFixative { get; init; }
    public string? EmbeddingMedium { get; init; }
    public SamplePreparation? SamplePreparation { get; init; }
    public SlideContainer? SlideContainer { get; init; }
}

/// <summary>Whole-slide-imaging data keyed by bioptic number.</summary>
public sealed record WsiData : Entity
{
    public required string BiopticNumber { get; init; }
    public required string SourceId { get; init; }
    public FixedBlock? FixedBlock { get; init; }
}
